controllers/rate_limiting_wasmplugin_controller.go

```diff
@@ -0,0 +1,369 @@
+	sort.Sort(common.PolicyByKey(rateLimitPoliciesSorted))
+
+	for _, policy := range rateLimitPolicies {
+		rlp := policy.(*kuadrantv1beta2.RateLimitPolicy)
+		wasmRLP, err := r.WASMRateLimitPolicy(ctx, gatewayAPITopology, rlp, gw)
+		if err != nil {
+			return nil, err
+		}
+
+		if wasmRLP == nil {
+			// skip this RLP
+			continue
+		}
+
+		wasmPlugin.RateLimitPolicies = append(wasmPlugin.RateLimitPolicies, *wasmRLP)
+	}
+
+	return wasmPlugin, nil
+}
+
+func (r *RateLimitingWASMPluginReconciler) gatewayAPITopologyFromGateway(ctx context.Context, gw *gatewayapiv1.Gateway) (*common.KuadrantTopology, error) {
+	logger, err := logr.FromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	routeList := &gatewayapiv1.HTTPRouteList{}
+	// Get all the routes having the gateway as parent
+	err = r.Client().List(ctx, routeList, client.MatchingFields{HTTPRouteParents: client.ObjectKeyFromObject(gw).String()})
+	logger.V(1).Info("gatewayAPITopologyFromGateway: list httproutes from gateway", "err", err)
+	if err != nil {
+		return nil, err
+	}
+
+	rlpList := &kuadrantv1beta2.RateLimitPolicyList{}
+	// Get all the rate limit policies
+	// TODO(eastizle): Add index field??
+	err = r.Client().List(ctx, rlpList)
+	logger.V(1).Info("gatewayAPITopologyFromGateway: list rate limit policies", "err", err)
+	if err != nil {
+		return nil, err
+	}
+
+	return common.NewKuadrantTopology(
+		[]*gatewayapiv1.Gateway{gw},
+		common.Map(routeList.Items, func(r gatewayapiv1.HTTPRoute) *gatewayapiv1.HTTPRoute { return &r }),
+		common.Map(rlpList.Items, func(p kuadrantv1beta2.RateLimitPolicy) common.KuadrantPolicy { return &p }),
+	), nil
+}
+
+func (r *RateLimitingWASMPluginReconciler) WASMRateLimitPolicy(ctx context.Context, t *common.KuadrantTopology, rlp *kuadrantv1beta2.RateLimitPolicy, gw *gatewayapiv1.Gateway) (*wasm.RateLimitPolicy, error) {
+	gwHostnamesTmp := common.TargetHostnames(gw)
+	gwHostnames := common.Map(gwHostnamesTmp, func(str string) gatewayapiv1.Hostname { return gatewayapiv1.Hostname(str) })
+
+	route, err := r.RouteFromRLP(ctx, t, rlp, gw)
+	if err != nil {
+		return nil, err
+	}
+	if route == nil {
+		// no need to add the policy if there are no routes;
+		// a rlp can return no rules if all its limits fail to match any route rule
+		// or targeting a gateway with no "free" routes. "free" meaning no route with policies targeting it
+		return nil, nil
+	}
+
+	rules := rlptools.WasmRules(rlp, route)
+	if len(rules) == 0 {
+		// no need to add the policy if there are no rules; a rlp can return no rules if all its limits fail to match any route rule
+		return nil, nil
+	}
+
+	// narrow the list of hostnames specified in the route so we don't generate wasm rules that only apply to other gateways
+	// this is a no-op for the gateway rlp
+	hostnames := common.FilterValidSubdomains(gwHostnames, route.Spec.Hostnames)
+	if len(hostnames) == 0 { // it should only happen when the route specifies no hostnames
+		hostnames = gwHostnames
+	}
```

Comment:
Shouldn't the wasm rules be computed _after_ fixing the hostnames for rules that don't declare any?

Reply:
good catch. When any route selector had not empty `hostnames` fields, the routes without hostname were skipped. Added e2e test to cover that use case as well.